pulseguardian/web.py

```diff
@@ -350,38 +351,79 @@ def update_info():
     pulse_username = request.form['pulse-user']
     new_password = request.form['new-password']
     password_verification = request.form['new-password-verification']
+    new_owners = _clean_owners_str(request.form['owners-list'])
 
     try:
         pulse_user = PulseUser.query.filter(
             PulseUser.username == pulse_username).one()
     except sqlalchemy.orm.exc.NoResultFound:
-        return profile(messages=["Invalid user."])
-
-    if pulse_user.owner != g.user:
-        return profile(messages=["Invalid user."])
-
-    if not new_password:
-        return profile(messages=["You didn't enter a new password."])
-
-    if new_password != password_verification:
-        return profile(error="Password verification doesn't match the "
-                       "password.")
-
-    if not PulseUser.strong_password(new_password):
-        return profile(error="Your password must contain a mix of "
-                       "letters and numerical characters and be at "
-                       "least 6 characters long.")
+        return profile(
+            messages=["Pulse user {} not found.".format(pulse_username)])
+
+    if g.user not in pulse_user.owners:
+        return profile(
+            messages=["Invalid user: {} is not an owner.".format(g.user.email)])
+
+    messages = []
+    error = None
+    if new_password:
+        if new_password != password_verification:
+            return profile(error="Password verification doesn't match the "
+                           "password.")
+
+        if not PulseUser.strong_password(new_password):
+            return profile(error="Your password must contain a mix of "
+                           "letters and numerical characters and be at "
+                           "least 6 characters long.")
+
+        pulse_user.change_password(new_password)
+        messages.append("Password updated for user {0}.".format(
+                        pulse_username))
+
+    # update the owners list, if needed
+    old_owners = {user.email for user in pulse_user.owners}
+    if new_owners and new_owners != old_owners:
+        # the list was changed.  Do an update
+        new_owner_users = list(User.query.filter(User.email.in_(new_owners)))
+        if new_owner_users:
+            # at least some of the new owners are real users in the db
+            pulse_user.owners = new_owner_users
+            db_session.commit()
+
+            updated_owners = {user.email for user in new_owner_users}
+            invalid_owners = list(new_owners - updated_owners)
+            invalid_owners.sort()
```

Comment:
A little cleaner to do

    invalid_owners = sorted(new_owners - updated_owners)

instead of those two lines (`sorted()` takes an iterable and returns a list).